packages/dds/tree/docs/wip/inverse-changes/README.md

```diff
@@ -0,0 +1,836 @@
+This means that with this approach,
+as far as this concurrency case is concerned,
+a client *can* get stuck repeatedly trying to undo a change.
+
+##### Update by Peer
+
+Peers receiving the undo are responsible for updating the undo change as appropriate.
+This means clients need to have access to repair data for changes within the collaboration window.
+
+#### Case #3: Changes Concurrent to and Sequenced After Both the Original and Undo
+
+Such changes need rebasing over
+the original change, any interim changes, and the undo change.
+This is something peers should be equipped to perform without need for additional data as part of the normal rebase process.
+
+It does mean however that we must design inverse changes and the output of rebase to ensure that
+a change being rebased over an inverse change
+after having been rebased over the original would undo the effects of rebasing over the original.
+For example, if the original change was a slice-delete and the rebased change was an insert that would commute with the slice,
+then the final rebased change should not be affected by the slice delete.
+This requires that inverse changes are able to counter the effects of
+the change they seek to counteract in changes that are sequenced after that inverse.
+See [Deriving an Inverse Changeset](#deriving-an-inverse-changeset) for more details.
+
+#### Data Sourcing Needs
+
+The client issuing the undo needs to have access to the following as far back as the undo window extends:
+* The original change to be undone (in its most rebased form to date) and its associated repair data
+* All of the interim changes sequenced since the change to be undone and their associated repair data
+
+In designs where peers accept the undo change as is
+(see "Looser Semantics" under
+[Case #1](#case-1-changes-concurrent-to-and-sequenced-before-the-original)
+and
+[Case #2](#case-2-changes-sequenced-after-the-original-but-concurrent-to-and-sequenced-before-the-undo)
+above),
+peers do not need any additional edit or repair data.
+
+In designs where peers are rejecting undo changesets that are stale
+(see "Validation By Peer" under
+[Case #1](#case-1-changes-concurrent-to-and-sequenced-before-the-original)
+and
+[Case #2](#case-2-changes-sequenced-after-the-original-but-concurrent-to-and-sequenced-before-the-undo)
+above),
+peers only need the relevant reference sequence number for the undo edit,
+which the Fluid sequencing service provides.
+
+In designs where peers perform additional computation to update (or re-derive) the undo,
+(see "Recomputation by Peer" under
+[Case #1](#case-1-changes-concurrent-to-and-sequenced-before-the-original)
+and "Update by Peer" under
+[Case #2](#case-2-changes-sequenced-after-the-original-but-concurrent-to-and-sequenced-before-the-undo)
+above),
+peers need to have access to the following as far back as the collaboration window extends:
+* Changes and their associated repair data
+
+#### Data Communication Needs
```

Comment:
Are you thinking of an approach where the changeset sent by the issuing client would not contain the repair data but instead contain some characterization of it, which peers would then be able to fill out? If so, do think of that as something different than what is described in the "Late Repair Data Concretization" section?

Reply:
For the guaranteed synchronous design, it could be what's described in Late Repair Data Concretization. I don't have a specific design proposal, but I think it would be nice if (assuming we're not sending all repair data inline) we always sent the references to the repair information in a format that makes it easy both to craft queries to a service and to find the information locally if it exists. Just sending a URL to a blob with all the repair data, for example, would not meet the criteria I'm proposing, because peers which had the relevant repair data available locally would probably not be able to convert that URL into a query on their local data.

Reply:
I added a "Abstract Repair Data Description" section here. I agree it's good to call out here. Documentation should favor clarity at the risk of redundancy.